<a href="https://colab.research.google.com/github/sijuswamy/Python-for-Datascience/blob/main/Recommender_rough_work_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Use the feature vectors for predicting the rating scores

Consider the problem with content based rating informations of 4 users for 6 movies. Construct a rating prediction using scalar projection.

$\begin{array}{|ll|l|r|r|r|r|r|r|}
\hline
{}&{}&F1&0.5&0.3&1&0.5&0&1\\
{}&{}&F2&0.7&0.9&0&0.9&1&0\\
\hline
F1&F2&{} &    M1 &    M2 &   M3 &    M4 &   M5 &   M6 \\
\hline
0.9&0.2&U1 &  &   &   &   &  &   \\\hline
0.4&0.9&U2 &   &   &   &   &   &   \\\hline
0.5&0.3&U3 &   &   &   &   &   &  \\\hline
0.2&0.7&U4 &   &   &   &   &   &   \\
\hline
\end{array}$

In [1]:
import numpy as np
import numpy.linalg as la

In [2]:
MF=np.array([[0.5,0.7],[0.3,0.9],[1,0],[0.5,0.9],[0,1],[1,0]])
UT=np.array([[0.9,0.2],[0.4,0.9],[0.5,0.3],[0.2,0.7]])
DP=np.dot(UT,MF.T)

In [3]:
print(DP)

[[0.59 0.45 0.9  0.63 0.2  0.9 ]
 [0.83 0.93 0.4  1.01 0.9  0.4 ]
 [0.46 0.42 0.5  0.52 0.3  0.5 ]
 [0.59 0.69 0.2  0.73 0.7  0.2 ]]


$\begin{array}{|ll|l|r|r|r|r|r|r|}
\hline
{}&{}&F1&0.5&0.3&1&0.5&0&1\\
{}&{}&F2&0.7&0.9&0&0.9&1&0\\
\hline
F1&F2&{} &    M1 &    M2 &   M3 &    M4 &   M5 &   M6 \\
\hline
0.9&0.2&U1 &  0.59 &  0.45 &  0.9 &  0.63 &  0.2 &  0.9 \\\hline
0.4&0.9&U2 &  0.83 &  0.93 &  0.4 &  1.01 &  0.9 &  0.4 \\\hline
0.5&0.3&U3 &  0.46 &  0.42 &  0.5 &  0.52 &  0.3 &  0.5 \\\hline
0.2&0.7&U4 &  0.59 &  0.69 &  0.2 &  0.73 &  0.7 &  0.2 \\
\hline
\end{array}$

In [4]:
import pandas as pd
DP1=pd.DataFrame(DP,columns=["M1","M2","M3","M4","M5","M6"],index=["U1","U2","U3","U4"])
print(DP1.to_latex())

\begin{tabular}{lrrrrrr}
\toprule
{} &    M1 &    M2 &   M3 &    M4 &   M5 &   M6 \\
\midrule
U1 &  0.59 &  0.45 &  0.9 &  0.63 &  0.2 &  0.9 \\
U2 &  0.83 &  0.93 &  0.4 &  1.01 &  0.9 &  0.4 \\
U3 &  0.46 &  0.42 &  0.5 &  0.52 &  0.3 &  0.5 \\
U4 &  0.59 &  0.69 &  0.2 &  0.73 &  0.7 &  0.2 \\
\bottomrule
\end{tabular}



$\begin{array}{lrrrrrr}
\hline
{} &    M1 &    M2 &   M3 &    M4 &   M5 &   M6 \\
\hline
U1 &  0.59 &  0.45 &  0.9 &  0.63 &  0.2 &  0.9 \\
U2 &  0.83 &  0.93 &  0.4 &  1.01 &  0.9 &  0.4 \\
U3 &  0.46 &  0.42 &  0.5 &  0.52 &  0.3 &  0.5 \\
U4 &  0.59 &  0.69 &  0.2 &  0.73 &  0.7 &  0.2 \\
\hline
\end{array}$

$\begin{array}{lrrrrrr}
\hline
{} &    M1 &    M2 &   M3 &    M4 &   M5 &   M6 \\
\hline
U1 &  0.59 &  0.45 &  0.9 &  0.63 &  0.2 &  0.9 \\
U2 &  0.83 &  0.93 &  0.4 &  1.01 &  0.9 &  0.4 \\
U3 &  0.46 &  0.42 &  0.5 &  0.52 &  0.3 &  0.5 \\
U4 &  0.59 &  0.69 &  0.2 &  0.73 &  0.7 &  0.2 \\
\hline
\end{array}$

## A samples Prediction Calculation Model for User based Colaborative recommender

Consider the data

$\begin{array}{lrrrrr}
\hline
{} &    M1 &    M2 &   M3 &    M4 &   M5  \\
\hline
Alice &  5 &  4 &  1 & 4 & ?  \\
N1 &  3 &  1 &  2 &  3 &  3  \\
N2 &  4 &  3 &  4&  3 &  5  \\
N3 &  3 &  3 &  1 &  5 &  4  \\
\hline
\end{array}$

Try to predict the rating score for $M5$ by Alice


## Rating data

In [6]:
Alice=np.array([5, 4, 1, 4])
N1=np.array([3,1,2,3])
N2=np.array([4,3,4,3])
N3=np.array([3,3,1,5])

## Calulating The Pearson similarity

$Sim(U,V)=\dfrac{\sum (r_{ui}-\bar{r_u})(r_{vi}-\bar{r_v})}{\sqrt{\sum (r_{ui}-\bar{r_u})^2}\sqrt{\sum (r_{vi}-\bar{r_v})^2}}$

In [7]:
mean_rating_Alice=np.mean(Alice)
User_means=np.array([np.mean(N1),np.mean(N2),np.mean(N3)])
Alice=Alice-np.mean(Alice)
N1=N1-np.mean(N1)
N2=N2-np.mean(N2)
N3=N3-np.mean(N3)

In [9]:
simA_N1=(np.dot(Alice,N1))/((la.norm(Alice))*la.norm(N1))

In [ ]:
print('Sim(Alice,N1):',simA_N1)

0.30151134457776363


In [11]:
simA_N2=(np.dot(Alice,N2))/((la.norm(Alice))*la.norm(N2))

In [ ]:
print('Sim(Alice,N2):',simA_N2)

Sim(Alice,N2): -0.3333333333333333


In [12]:
simA_N3=(np.dot(Alice,N3))/((la.norm(Alice))*la.norm(N3))

In [ ]:
print('Sim(Alice,N3):',simA_N3)

Sim(Alice,N3): 0.7071067811865475


## Calculation of Rating for $M5$ by Alice.

\begin{equation*}
        P(Alice,M5)=\bar{r}_{Alice}+\dfrac{\sum\limits_{i}sim(Alice,N_i)\left(r(N_i,M5)-\bar{r}_{N_i}\right)}{\sum\limits_{i}|sim(Alice,N_i)|}
    \end{equation*}

In [13]:
RM5=np.array([3,5,4])
simarray=np.array([simA_N1,simA_N2,simA_N3])
meandiff=RM5-User_means
RUserF=np.dot(simarray,meandiff)/np.sum(abs(simarray))

In [14]:
r_alice_M5=mean_rating_Alice+RUserF
print('Rating for Movie 5 by Alice is:',r_alice_M5)

Rating for Movie 5 by Alice is: 3.8228434878793527


## Eigen values


In [ ]:
import numpy as np
import numpy.linalg as la

In [ ]:
A=np.array([[1 ,2, 4],[2,5,6],[1,9,7]])
ev,eval=la.eig(A)
ev


array([14.14537808+0.j        , -0.57268904+0.93479048j,
       -0.57268904-0.93479048j])